src/lib.rs

```diff
@@ -0,0 +1,386 @@
+    // and not worrying too much about performance.
+    pub fn vals(&self, vals: &[Val], pairs: &mut Pairs) -> Result<()> {
+        let result = unsafe {
+            ffi::rill_query_vals(self.query, vals.as_ptr(), vals.len(), pairs.pairs)
+        };
+        if result.is_null() { return err(); }
+
+        pairs.pairs = result;
+        Ok(())
+    }
+}
+
+impl Drop for Query {
+    fn drop(&mut self) {
+        unsafe { ffi::rill_query_close(self.query) }
+    }
+}
+
+#[test]
+fn test_rotate_query() {
+    let dir = Path::new("/tmp/rill-rs.rotate.test");
+    let _ = std::fs::remove_dir_all(dir);
+
+    {
+        let mut acc = Acc::new(dir, 2).unwrap();
+
+        acc.ingest(1, 10);
+        rotate(dir, 1 * 60 * 60).unwrap();
+
+        acc.ingest(2, 10);
+        acc.ingest(1, 10);
+        rotate(dir, 2 * 60 * 60).unwrap();
+
+        acc.ingest(2, 10);
+        acc.ingest(1, 20);
+        rotate(dir, 3 * 60 * 60).unwrap();
+
+        acc.ingest(1, 30);
+    }
+
+    let query = Query::new(dir).unwrap();
+
+    {
+        let mut pairs = Pairs::with_capacity(1).unwrap();
+        query.key(2, &mut pairs).unwrap();
+        assert_eq!(pairs.len(), 1);
+        assert_eq!(*pairs.get(0), KV{key: 2, val: 10});
+    }
+
+    {
+        let mut pairs = Pairs::with_capacity(1).unwrap();
+        query.keys(&[1], &mut pairs).unwrap();
+        assert_eq!(pairs.len(), 2);
+        assert_eq!(*pairs.get(0), KV{key: 1, val: 10});
+        assert_eq!(*pairs.get(1), KV{key: 1, val: 20});
+    }
+
+    {
+        let mut pairs = Pairs::with_capacity(1).unwrap();
+        query.keys(&[2, 3], &mut pairs).unwrap();
+        assert_eq!(pairs.len(), 1);
+        assert_eq!(*pairs.get(0), KV{key: 2, val: 10});
+    }
+
+    {
+        let mut pairs = Pairs::with_capacity(1).unwrap();
+        query.vals(&[10], &mut pairs).unwrap();
+        assert_eq!(pairs.len(), 2);
+        assert_eq!(*pairs.get(0), KV{key: 1, val: 10});
+        assert_eq!(*pairs.get(1), KV{key: 2, val: 10});
+    }
+
+    {
+        let mut pairs = Pairs::with_capacity(1).unwrap();
+        query.vals(&[20, 30], &mut pairs).unwrap();
+        assert_eq!(pairs.len(), 1);
+        assert_eq!(*pairs.get(0), KV{key: 1, val: 20});
+    }
+}
+
+pub struct Store {
+    store: *mut ffi::rill_store
+}
+
+impl Store {
+    pub fn open(dir: &Path) -> Result<Store> {
+        let c_dir = path_to_c_str(dir)?;
+        let store = unsafe { ffi::rill_store_open(c_dir.as_ptr()) };
+        if !store.is_null() { Ok(Store{store: store}) } else { err() }
+    }
+
+    pub fn write(dir: &Path, ts: Ts, quant: usize, pairs: &Pairs) -> Result<()> {
+        let c_dir = path_to_c_str(dir)?;
+        let ret = unsafe { ffi::rill_store_write(c_dir.as_ptr(), ts, quant, pairs.pairs) };
+        if ret { Ok(()) } else { err() }
+    }
+
+    pub fn merge(out: &Path, ts: Ts, quant: usize, stores: &[Store]) -> Result<()> {
+        let c_out = path_to_c_str(out)?;
+        let c_list: Vec<*const ffi::rill_store> =
```

Comment:
Maybe `Vec<_>` is enough, letting the item type to be inferred by the compiler.

Reply:
If I remember correctly it infered a `*mut` which caused problems later on so I had to force a `*const` here to avoid more crap later on. I may have done something wrong elsewhere as well.

Reply:
Just retested it and `Vec<_>` works fine so nevermind.